zest/releaser/vcs.py

```diff
@@ -204,33 +246,41 @@ def _update_version(self, version):
             setup_version = self.get_setup_py_version()
             if not setup_version or (setup_version ==
                                      self.get_version_txt_version()):
-                open(versionfile, 'w').write(version + '\n')
-                logger.info("Changed %s to %r", versionfile, version)
+                open(versionfile, 'w').write(version + u'\n')
+                logger.info(
+                    u"Changed {0} to {1!r}".format(versionfile, version)
+                    )
                 return
 
-        good_version = "version = '%s'" % version
-        line_number = 0
-        setup_lines = open('setup.py').read().split('\n')
-        for line_number, line in enumerate(setup_lines):
-            match = VERSION_PATTERN.search(line)
-            if match:
-                logger.debug("Matching version line found: %r", line)
-                if line.startswith(' '):
-                    # oh, probably '    version = 1.0,' line.
-                    indentation = line.split('version')[0]
-                    # Note: no spaces around the '='.
-                    good_version = indentation + "version='%s'," % version
-                setup_lines[line_number] = good_version
-                contents = '\n'.join(setup_lines)
-                open('setup.py', 'w').write(contents)
-                logger.info("Set setup.py's version to %r", version)
-                return
+        content = read_text_file(filename)
+        tree = ast.parse(content, filename)
+        assignment = self._find_assignment(tree, u'version')
+        if assignment:
+            # This assumes that the version string does not span more than
+            # one line
+            position = (
+                assignment.value.lineno - 1,
+                assignment.value.col_offset,
+                assignment.value.col_offset + len(repr(assignment.value.s))
+                )
+            lines = content.splitlines()
+            logger.debug(u"Matching version line found: {0!r}".format(
+                lines[position[0]]))
+            lines[position[0]] = (
+                lines[position[0]][:position[1]] +
+                repr(version) +
+                lines[position[0]][position[2]:]
+                )
+            contents = u'\n'.join(lines)
+            open('setup.py', 'w').write(contents)
+            logger.info("Set setup.py's version to {0}".format(version))
+            return
```

Comment:
Mitchell: don't put too much effort into it yet. I get the feeling that you're trying to make this piece of code very efficient.

The end result is code that is, on first sight, hardly readable for me. "position", "col_offset"... To me, it looks like it is hard to maintain that way. A good old simple `lines[...] = good_version` is much more readable.

Perhaps AST parsing is fine, but a core principle of zest.releaser, to me, has always been to keep it as simple as possible. That way, exceptions are easier to handle. 

So: I don't mind a bit of brute-force iterating and a bit of regex'ing. It doesn't need to be performant, it just needs to work. And I especially should be able to read it in 2018 to fix some weird bug. So readability counts much more than a bit of purity or speed.

Just wanted to say that before you pour in even more changes into this super-big pull request :-)